### PR TITLE
[BEAM-3419] Explicitly model iterable side inputs as part of the Beam portability APIs.

### PR DESCRIPTION
**Pull request by lukecwik**
Also clarify that accessing the "iterable" side input for a multimap side input will enumerate the KV pairs.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by lukecwik** [comment]
R: @youngoli

**Comment by lukecwik** [comment]
CC: @robertwb

**Comment by lukecwik** [comment]
Run Java PreCommit

File: model/fn-execution/src/main/proto/beam_fn_api.proto [hunk 1]
```diff
@@ -661,11 +668,28 @@ message StateKey {
     bytes key = 4;
   }
 
+  // Represents a request for the values associated with a specified window in a PCollection.
+  //
+  // Can only perform StateGetRequests on side inputs with the URN beam:side_input:iterable:v1 and
```
Comment (youngoli) [inline model/fn-execution/src/main/proto/beam_fn_api.proto]: Ditto with the comment above.

File: model/pipeline/src/main/proto/beam_runner_api.proto [hunk 2]
```diff
@@ -318,7 +318,15 @@ message StandardPTransforms {
...
 message StandardSideInputTypes {
   enum Enum {
+    // Represents a view over a PCollection<V>.
+    //
+    // The SDK is limited to perform state get requests using the StateKey.IterableSideInput.
     ITERABLE = 0 [(beam_urn) = "beam:side_input:iterable:v1"];
+
+    // Represents a view over a PCollection<KV<K, V>>.
+    //
+    // The SDK is able to perform state get requests with the StateKey.IterableSideInput and
```
Comment (youngoli) [inline model/pipeline/src/main/proto/beam_runner_api.proto]: In the same vein as the comments above, I think these lines would be less ambiguous like so:

"StateGetRequests performed on this side input must use StateKey.IterableSideInput or StateKey.MultimapSideInput."

Like my other comments, the important change is that StateGetRequests are explicitly mentioned as _using_ the StateKey and being performed _on_ the side input represented by this URN.

File: model/fn-execution/src/main/proto/beam_fn_api.proto [hunk 3]
```diff
@@ -637,6 +637,13 @@ message StateKey {
     bytes key = 1;
   }
 
+  // Represents a request for the values associated with a specified user key and window
+  // in a PCollection.
+  //
+  // Can only perform StateGetRequests on side inputs with the URN beam:side_input:multimap:v1.
```
[the remark this answers is not in the document]
Comment (lukecwik) [reply model/fn-execution/src/main/proto/beam_fn_api.proto]: Done

File: model/fn-execution/src/main/proto/beam_fn_api.proto (hunk 1 above)
Comment (lukecwik) [reply model/fn-execution/src/main/proto/beam_fn_api.proto]: Done

**Comment by lukecwik** [comment]
Thanks for the comments.

**Comment by lukecwik** [comment]
Run Java PreCommit

**Comment by lukecwik** [comment]
Run Python PreCommit

**Pull request merged**